src/pcrlock/pehash.c

```diff
@@ -0,0 +1,520 @@
+        le64_t SizeOfHeapCommit;
+        le32_t LoaderFlags;
+        le32_t NumberOfRvaAndSizes;
+        IMAGE_DATA_DIRECTORY DataDirectory[];
+} IMAGE_OPTIONAL_HEADER_64;
+
+typedef union IMAGE_OPTIONAL_HEADER_UNION {
+        IMAGE_OPTIONAL_HEADER_32 o32;
+        IMAGE_OPTIONAL_HEADER_64 o64;
+        le16_t Magic;
+} IMAGE_OPTIONAL_HEADER_UNION;
+
+typedef struct PeHeader {
+        le32_t signature;
+        IMAGE_FILE_HEADER pe;
+        IMAGE_OPTIONAL_HEADER_UNION optional;
+} PeHeader;
+
+typedef struct _IMAGE_SECTION_HEADER {
+        uint8_t Name[8];
+        le32_t VirtualSize;
+        le32_t VirtualAddress;
+        le32_t SizeOfRawData;
+        le32_t PointerToRawData;
+        le32_t PointerToRelocations;
+        le32_t PointerToLinenumbers;
+        le16_t NumberOfRelocations;
+        le16_t NumberOfLinenumbers;
+        le32_t Characteristics;
+} IMAGE_SECTION_HEADER;
+
+#define IMAGE_DATA_DIRECTORY_INDEX_CERTIFICATION_TABLE 4U
+
+static bool pe_header_is_64bit(const PeHeader *h) {
+        assert(h);
+
+        if (h->optional.Magic == UINT16_C(0x010B)) /* PE32 */
+                return false;
+
+        if (h->optional.Magic == UINT16_C(0x020B)) /* PE32+ */
+                return true;
+
+        assert_not_reached();
+}
+
+#define PE_HEADER_OPTIONAL_FIELD(h, field)                           \
+        (*(pe_header_is_64bit(h) ? &(h)->optional.o64.field : &(h)->optional.o32.field))
+
+#define PE_HEADER_OPTIONAL_FIELD_OFFSET(h, field) \
+        (pe_header_is_64bit(h) ? offsetof(PeHeader, optional.o64.field) : offsetof(PeHeader, optional.o32.field))
+
+static IMAGE_DATA_DIRECTORY *pe_header_get_data_directory(PeHeader *h, size_t i) {
+        assert(h);
+
+        if (i >= PE_HEADER_OPTIONAL_FIELD(h, NumberOfRvaAndSizes))
+                return NULL;
+
+        return PE_HEADER_OPTIONAL_FIELD(h, DataDirectory) + i;
+}
+
+static int hash_file(int fd, EVP_MD_CTX *md_ctx, uint64_t offset, uint64_t size) {
+        uint8_t buffer[64*1024];
+
+        log_debug("Hashing %" PRIu64 " @ %" PRIu64 " → %" PRIu64, size, offset, offset + size);
+
+        while (size > 0) {
+                size_t m = MIN(size, sizeof(buffer));
+                ssize_t n;
+
+                n = pread(fd, buffer, m, offset);
+                if (n < 0)
+                        return log_debug_errno(errno, "Failed to read file for hashing: %m");
+                if ((size_t) n != m)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG), "Short read while hashing.");
+
+                if (EVP_DigestUpdate(md_ctx, buffer, m) != 1)
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE), "Unable to hash data.");
+
+                offset += m;
+                size -= m;
+        }
+
+        return 0;
+}
+
+static int section_offset_cmp(const IMAGE_SECTION_HEADER *a, const IMAGE_SECTION_HEADER *b) {
+        return CMP(a->PointerToRawData, b->PointerToRawData);
+}
+
+static int pe_load_headers(
```

Comment:
I believe this is our 3nd PE parser in userspace, 4th if you count the one in efi space. You might wanna factor that stuff out… :P

Reply:
→ #28955